### PR TITLE
ERC5792 experimental

### PR DESCRIPTION
**Pull request by KannuSingh**
changes in react-dapp-v2, added button to test wallet_getCapabilities and wallet_sendCalls rpc method. This can be used to test it.
changes in  react-wallet-v2, added support in the wallet to handle new rpc method (wallet_getCapabilities, wallet_sendCalls)

Todo: check for requesting account type(EOA or SCA) and handle accordingly, currently assuming its a smart contract account.

### CODE REVIEW

File: advanced/dapps/react-dapp-v2/src/helpers/tx.ts [hunk 1]
```diff
@@ -41,3 +43,32 @@ export async function formatTestTransaction(account: string) {
 
   return tx;
 }
+
+
+export async function formatTestBatchCall(account: string) {
+  const [namespace, reference, address] = account.split(":");
+  // calldata for batch send
+  const receiverAddress = '0xc3cE257B5e2A2ad92747dd486B38d7b4B36Ac7C9'
```
Comment (lukaisailovic) [inline advanced/dapps/react-dapp-v2/src/helpers/tx.ts:51]: What is this address?

File: advanced/dapps/react-dapp-v2/src/helpers/tx.ts [hunk 2]
```diff
@@ -41,3 +43,32 @@ export async function formatTestTransaction(account: string) {
...
+      data:'0x' as `0x${string}`,
+      value: amountToSend as `0x${string}`
+    },
+  ]
+  const sendCallsRequestParams:SendCallsParams = {
+    version:'1.0',
+    chainId: `0x${BigInt(reference).toString(16)}`,
+    from: address as `0x${string}`,
+    calls: calls,
+    // capabilities: { paymasterService: { url: 'http://localhost:3002/api' } },
```
Comment (lukaisailovic) [inline advanced/dapps/react-dapp-v2/src/helpers/tx.ts:70]: Remove comment

File: advanced/wallets/react-wallet-v2/src/hooks/useWalletConnectEventsManager.ts [hunk 3]
```diff
@@ -41,12 +46,13 @@ export default function useWalletConnectEventsManager(initialized: boolean) {
...
       console.log('session_request', requestEvent)
-      const { topic, params, verifyContext } = requestEvent
+      const { topic, params, verifyContext, id } = requestEvent
       const { request } = params
       const requestSession = web3wallet.engine.signClient.session.get(topic)
       // set the verify context so it can be displayed in the projectInfoCard
       SettingsStore.setCurrentRequestVerifyContext(verifyContext)
-
+      const wallet = await getWallet(params)
+      console.log("Got request for : ",request.method)
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/hooks/useWalletConnectEventsManager.ts:55]: remove log

File: advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  switch (request.method) {
+    case EIP5792_METHODS.WALLET_GET_CAPABILITIES:
+      /**
+       * If the account in the params is a EOA return supportedEIP5792CapabilitiesForEOA
+       * If the account in the params is a Smart Contract return supportedEIP5792CapabilitiesForSCA
+       */
+      if(wallet instanceof EIP155Lib)
+        return formatJsonRpcResult<GetCapabilitiesResult>(id, supportedEIP5792CapabilitiesForEOA)
+      else if(wallet instanceof SmartAccountLib || wallet instanceof KernelSmartAccountLib)
+        return formatJsonRpcResult<GetCapabilitiesResult>(id, supportedEIP5792CapabilitiesForSCA)
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts:35]: if we return we don't need else

Comment (lukaisailovic) [reply advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts:35]: We may also want  to define some (maybe even empty?) interface that will be implemented by `SmartAccountLib` and `KernelSmartAccountLib` so we can just check for that

File: advanced/wallets/react-wallet-v2/src/lib/smart-accounts/SmartAccountLib.ts [hunk 5]
```diff
@@ -184,4 +184,49 @@ export abstract class SmartAccountLib implements EIP155Wallet {
...
+    // userOp.signature = newSignature
+
+    // const userOpHash = await this.bundlerClient.sendUserOperation({
+    // userOperation: userOp
+    // })
+    // let userOpsReceipt = await this.bundlerClient.waitForUserOperationReceipt({
+    //   hash: userOpHash
+    // });
+    // return userOpsReceipt.receipt.transactionHash;
+  }
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/lib/smart-accounts/SmartAccountLib.ts:231]: Lets remove the comments

File: advanced/wallets/react-wallet-v2/src/utils/EIP155WalletUtil.ts [hunk 6]
```diff
@@ -41,3 +43,48 @@ export function createOrRestoreEIP155Wallet() {
...
+  if (eoaWallet) {
+    return eoaWallet
+  }
+
+  /**
+   * Smart accounts
+   */
+  const chainId = params?.chainId?.split(':')[1]
+  console.log('Chain ID', { chainId })
+  console.log('PARAMS', { params })
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/utils/EIP155WalletUtil.ts:61]: Remove comments

File: advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import EIP155Lib from '@/lib/EIP155Lib'
+import { SmartAccountLib } from '@/lib/smart-accounts/SmartAccountLib'
+type RequestEventArgs = Omit<SignClientTypes.EventArguments['session_request'], 'verifyContext'>
+
+
+export async function approveEIP5792Request(requestEvent: RequestEventArgs) {
+  const { params, id } = requestEvent
+  const { chainId, request } = params
+
+  console.log(requestEvent, chainId, 'tests')
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts:20]: remove logs

File: advanced/wallets/react-wallet-v2/src/utils/EIP155WalletUtil.ts [hunk 8]
```diff
@@ -41,3 +43,48 @@ export function createOrRestoreEIP155Wallet() {
...
+  const lib = smartAccountWallets[`${chainId}:${address}`]
+  if (lib) {
+    return lib
+  }
+  console.log('Library not found', {
+    target: `${chainId}:address`,
+    values: Object.keys(smartAccountWallets)
+  })
+  throw new Error('Cannot find wallet for requested address')
+}
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/utils/EIP155WalletUtil.ts:90]: This code is copy pasted from `EIP155RequestHandlerUtil.ts`. If we extract it to different place, lets make sure to  use `getWallet` reference in the initial implementation

File: advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts (hunk 4 above)
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts:35]: Unnecessary else

**Review by lukaisailovic — COMMENTED** [review]
```
instead of sending the transaction and waiting for transaction receipt
I think it should submit the userOperation and return the userOperation hash,
Might need to add getTransactionStatus(userOpHash) to get the status,
might need to add getReceipt(userOpsHash) method to get the receipt.
```

We can add this in separate PR

File: advanced/dapps/react-dapp-v2/src/helpers/tx.ts (hunk 1 above)
Comment (KannuSingh) [reply advanced/dapps/react-dapp-v2/src/helpers/tx.ts:51]: Added my EOA address to test send ETH to it, can change it to something else.

File: advanced/wallets/react-wallet-v2/src/utils/EIP155WalletUtil.ts (hunk 8 above)
Comment (KannuSingh) [reply advanced/wallets/react-wallet-v2/src/utils/EIP155WalletUtil.ts:90]: sure, will be change it in the initial reference implementation to use this.

File: advanced/dapps/react-dapp-v2/src/helpers/tx.ts (hunk 1 above)
Comment (lukaisailovic) [reply advanced/dapps/react-dapp-v2/src/helpers/tx.ts:51]: Yes lets change it to something else. Good practice is to add the address of Sepolia Faucet if we are sending Sepolia ETH

File: advanced/dapps/react-dapp-v2/src/constants/default.ts [hunk 9]
```diff
@@ -68,6 +99,12 @@ export enum DEFAULT_EIP155_OPTIONAL_METHODS {
   ETH_SIGN = "eth_sign",
   ETH_SIGN_TYPED_DATA = "eth_signTypedData",
   ETH_SIGN_TYPED_DATA_V4 = "eth_signTypedData_v4",
+  /**
+   * EIP5792 methods
+   * Adding it here so when session is requested following permission are also requested
+   */
+  WALLET_GET_CAPABILITIES = "wallet_getCapabilities",
+  WALLET_SEND_CALLS = "wallet_sendCalls"
```
[the remark this answers is not in the document]
Comment (KannuSingh) [reply advanced/dapps/react-dapp-v2/src/constants/default.ts:107]: I was not sure whether to separated it based on EIP155 and EIP5792 in the dapp side, also when creating initial proposal request to wallet, it should consider both EIP155 Optional methods and EIP5792 method. It needed to change some existing method type, so decided to put it here instead. 
I will do the change such it is only defined at one place.

**Comment by KannuSingh** [comment]
> ```
> instead of sending the transaction and waiting for transaction receipt
> I think it should submit the userOperation and return the userOperation hash,
> Might need to add getTransactionStatus(userOpHash) to get the status,
> might need to add getReceipt(userOpsHash) method to get the receipt.
> ```
> 
> We can add this in separate PR

Other rpc methods wallet_getCallsStatus and wallet_showCallsStatus be also be part of  that separate PR?

File: advanced/wallets/react-wallet-v2/src/lib/smart-accounts/KernelSmartAccountLib.ts [hunk 10]
```diff
@@ -163,6 +163,34 @@ export class KernelSmartAccountLib implements EIP155Wallet {
...
+    const newSignature = await this.client.account.signUserOperation(userOp)
+    console.log('Signatures',{old: userOp.signature, new: newSignature});
+
+    userOp.signature = newSignature
+
+    const userOpHash = await this.client.sendUserOperation({
+    userOperation: userOp,
+    account: this.client.account
+    })
+    return userOpHash;
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/lib/smart-accounts/KernelSmartAccountLib.ts:192]: Why are we using fixed `preVerificationGas`?

File: advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        }))
+        const userOpHash = await wallet.sendBatchTransaction(sendCalls)
+        return formatJsonRpcResult(id, userOpHash)
+      } catch (error: any) {
+        console.error(error)
+        alert(error.message)
+        return formatJsonRpcError(id, error.message)
+      }
+    }
+    default:
```
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts:100]: Can we extract this code to their own methods within the same file?

File: advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts (hunk 4 above)
Comment (lukaisailovic) [inline advanced/wallets/react-wallet-v2/src/utils/EIP5792RequestHandlerUtils.ts]: No need for else

**Review by lukaisailovic — APPROVED** [review]

**Pull request merged**